### PR TITLE
Submit/external command tests

### PR DESCRIPTION
**Pull request by dcbaker**
Here's some tests for the ExternalCommand class. These don't hit everything, but they hit a lot of the class. These were on a branch that I had trying to using call_cmd and call_cmd_async for handling the calling that I never got working, but the tests themselves should be useful now that I've updated them to apply to master.

There's also a cleanup for a comment in here, and a patch to simplify some if branching, but it's not that complicated and the tests cover that change.

### CODE REVIEW

File: tests/commands/global_test.py [hunk 1]
```diff
@@ -147,3 +148,33 @@ def test_no_spawn_failure(self):
...
+                refocus=False, spawn=True)
+        cmd.apply(ui)
+        ui.notify.assert_not_called()
+
+    @mock.patch('alot.commands.globals.settings.get', mock.Mock(return_value=''))
+    def test_spawn_stdin_success(self):
+        ui = mock.Mock()
+        with mock.patch.dict(os.environ, {'DISPLAY': ':0'}):
+            cmd = g_commands.ExternalCommand(
+                u'python -c "import sys; sys.exit(int(sys.stdin.read().strip()))"',
```
Comment (lucc) [inline tests/commands/global_test.py]: `awk '{ exit $0 }'` takes 1/3 of python2, if you don't like the dependency on awk, `sh -c 'exit $(cat)'` takes 1/2 of python2.

File: tests/commands/global_test.py [hunk 2]
```diff
@@ -147,3 +148,33 @@ def test_no_spawn_failure(self):
...
+                stdin=u'0', refocus=False, spawn=True)
+        cmd.apply(ui)
+        ui.notify.assert_not_called()
+
+    @mock.patch('alot.commands.globals.settings.get', mock.Mock(return_value=''))
+    def test_spawn_failure(self):
+        ui = mock.Mock()
+        with mock.patch.dict(os.environ, {'DISPLAY': ':0'}):
+            cmd = g_commands.ExternalCommand(
+                u'python -c "import sys; sys.exit(1)"',
```
Comment (lucc) [inline tests/commands/global_test.py]: `false` is faster again.

File: tests/commands/global_test.py [hunk 3]
```diff
@@ -116,3 +117,64 @@ def test_apply_sign_by_default_but_no_key(self):
...
+
+    def test_no_spawn_failure(self):
+        ui = mock.Mock()
+        cmd = g_commands.ExternalCommand(
+            u'python -c "import sys; sys.exit(1)"',
+            refocus=False, on_success=self.on_success)
+        cmd.apply(ui)
+        ui.notify.assert_called_once_with('', priority='error')
+
+    @mock.patch('alot.commands.globals.settings.get', mock.Mock(return_value=''))
```
Comment (lucc) [inline tests/commands/global_test.py]: Is there a reason why you patch `settings.get` in a decorator and `os.environ` in a with block?

File: tests/commands/global_test.py [hunk 4]
```diff
@@ -116,3 +117,64 @@ def test_apply_sign_by_default_but_no_key(self):
...
+
+    class Success(Exception):
+        pass
+
+    def on_success(self):
+        raise self.Success
+
+    def test_no_spawn_no_stdin_success(self):
+        cmd = g_commands.ExternalCommand(
+            u'python -c "import sys; sys.exit(0)"',
```
Comment (lucc) [inline tests/commands/global_test.py]: speed, see below

File: tests/commands/global_test.py [hunk 5]
```diff
@@ -116,3 +117,64 @@ def test_apply_sign_by_default_but_no_key(self):
...
+            refocus=False, on_success=self.on_success)
+        with self.assertRaises(self.Success):
+            cmd.apply(mock.Mock())
+
+    def test_no_spawn_stdin_success(self):
+        cmd = g_commands.ExternalCommand(
+            u'python -c "import sys; sys.exit(int(sys.stdin.read().strip()))"',
```
Comment (lucc) [inline tests/commands/global_test.py]: speed ...

File: tests/commands/global_test.py [hunk 6]
```diff
@@ -116,3 +117,64 @@ def test_apply_sign_by_default_but_no_key(self):
...
+            stdin=u'0', refocus=False, on_success=self.on_success)
+        with self.assertRaises(self.Success):
+            cmd.apply(mock.Mock())
+
+    def test_no_spawn_failure(self):
+        ui = mock.Mock()
+        cmd = g_commands.ExternalCommand(
+            u'python -c "import sys; sys.exit(1)"',
```
Comment (lucc) [inline tests/commands/global_test.py]: speed ...

**Comment by dcbaker** [comment]
Updated to address lucc's comments.

**Comment by pazz** [comment]
This series introduces a strange and unexpected behaviour at my end:
When I reply to a mail and my editor is takes over the terminal, it seems that the keypresses do not reach the editor (vim).

I'm on 3ad6ebf3, the first commit which shows this behaviour.
The alot log sais this:
```
INFO:envelope:editable headers: set([u'Subject'])
DEBUG:globals:using editor_cmd: /usr/bin/vim -f -c 'set filetype=mail' +
DEBUG:globals:{'spawn: ': False, 'in_thread': False}
DEBUG:globals:{'spawn': False}
DEBUG:globals:cmdlist: ['/usr/bin/vim', '-f', '-c', 'set filetype=mail', '+', '/tmp/alot.hZVNjZ.eml']
INFO:globals:calling external command: ['/usr/bin/vim', '-f', '-c', 'set filetype=mail', '+', '/tmp/alot.hZVNjZ.eml']
```

My alot config contains this as the only editor-related setting (all others are commented out):
```
editor_cmd = /usr/bin/vim -f -c 'set filetype=mail' +
```

**Pull request merged**